Add Smart Snack Dispenser codelab (#63)
viam-devrel/viamcodelabs: site/viamguides/src/smart-snack-dispenser/smart-snack-dispenser.md

1. Check **Set hostname** and enter the name you would like to access the Pi by in that field, for example, `test`.
1. Select the checkbox for **Set username and password** and set a username (for example, your first name) that you will use to log into the Pi. If you skip this step, the default username will be `pi` (not recommended for security reasons). And specify a password.
1. Connect your Pi to Wi-Fi so that you can run `viam-server` wirelessly. Check **Configure wireless LAN** and enter your wireless network credentials. SSID (short for Service Set Identifier) is your Wi-Fi network name, and password is the network password. Change the section `Wireless LAN country` to where your router is currently being operated.
   ![raspberry pi hostname username and password](assets/generalSettings.webp)
1. Select the **SERVICES** tab, check **Enable SSH**, and select **Use password authentication**.
   ![raspberry pi enable SSH](assets/sshSettings.webp)
   > aside negative
   > Be sure that you remember the `hostname` and `username` you set, as you will need this when you SSH into your Pi.

> 1. **Save** your updates, and confirm `YES` to apply OS customization settings. Confirm `YES` to erase data on the USB flash drive. You may also be prompted by your operating system to enter an administrator password. After granting permissions to the Imager, it will begin writing and then verifying the Linux installation to the USB flash drive.

Reviewer: ```suggestion
1. **Save** your updates, and confirm `YES` to apply OS customization settings. Confirm `YES` to erase data on your storage device. You may also be prompted by your operating system to enter an administrator password. After granting permissions to the Imager, it will begin writing and then verifying the Linux installation to your storage device.
```
Accounts for use of USB or microSD card instead of having to spell out both.

> 1. Remove the USB flash drive from your computer when the installation is complete.

Reviewer: ```suggestion
1. Remove the storage device from your computer when the installation is complete.
```
Accounts for use of USB or microSD card instead of having to spell out both.

   ![raspberry pi SSH login](assets/sshLogin.webp)
1. Update your Raspberry Pi to ensure all the latest packages are installed
   ```bash
   sudo apt update
   sudo apt upgrade
   ```

Now that you've confirmed the Raspberry Pi is set up with an operating system, you can shut it down with the following command in the SSH session on the Pi:
```bash
sudo shutdown -h now
```

The SSH session will close as the Raspberry Pi begins to turn off.
Once the green activity light stops blinking, you can unplug the Raspberry Pi from power to prepare assembling the hardware in the next step.

<!-- ------------------------ -->
## Set up the hardware
Duration: 20

It's time to get your hands dirty by taking apart the food dispenser and upgrading the internals with the Raspberry Pi and other peripherals!

### Disassemble the snack dispenser

1. Remove the food bowl from the base of the dispenser and set it aside. 
   ![food dispenser without bowl](assets/hardwarePrepped.webp)
1. Turn the dispenser upside down to access the screws on the bottom of the device.
   ![flip the machine](assets/hardwareFlip.webp)

> 1. Use the screwdriver to remove the screws in each corner. Set them aside for in a safe place for later when reassembling.

Reviewer: ```suggestion
1. Use the screwdriver to remove the screws in each corner. Set them aside in a safe place for later when reassembling.
```

1. Set the food container part aside now that it is disconnected from the base. Place the base on its front so the back of the device is facing up.
   Put on safety glasses and use the drill to slowly bore a hole in the base a couple inches from the bottom. Make sure you have a secure grip to prevent it from spinning while drilling. It should enter the base just above the power port.
   ![camera cable hole](assets/hardwareCameraCableHole.webp)
1. This hole will be used by the webcam's USB cable to connect to the Raspberry Pi once it is embedded in the food dispenser. Test it by pushing the USB end through several inches.
   If there are any issues, use the drill to widen the entry point with a bigger drill bit or by slowly rotating the existing bit around the inside of the hole while the drill is running, applying light pressure to strip away material.
   ![test camera entry](assets/hardwareCameraCableEntry.webp)
1. Clean out any plastic material from the drilling process from the inside of the base and place the base with the bottom down to prepare for wiring.

### Connect hardware components

![wiring diagram](assets/wiring.webp)

> 1. Use the above wiring diagram to connect the hardware peripherals and power together. The stripped wires from the snack dispenser can be connected to the jumper wires unless the WAGO lever nuts.

Reviewer: I keep rereading `can be connected to the jumper wires unless the WAGO lever nuts`, but am unsure what that means 😅

Author: I meant to say "using" the WAGO lever nuts 🤦

1. Click the **+** icon in the left-hand menu and select **Component**.
   ![add component](assets/addComponent.webp)
1. Select `board`, and find the `raspberry-pi:rpi4` module. This adds the module for working with the Raspberry Pi 4's GPIO pins. Leave the default name `board-1` for now.
1. Notice adding this module adds the board hardware component called `board-1`. The collapsible card on the right corresponds to the part listed in the left sidebar.
   ![added board](assets/addBoard.webp)
1. Click **Save** in the top right to save and apply your configuration changes.

### Configure your encoder

1. In [the Viam app](https://app.viam.com/fleet/locations), find the **CONFIGURE** tab.
1. Click the **+** icon in the left-hand menu and select **Component**.
1. Select encoder, and find the `single` module. This adds the module for reading the limit switch as an encoder with a board's GPIO pins. Leave the default name `encoder-1` for now.
1. Notice adding this module adds the encoder hardware component called `encoder-1`. The collapsible card on the right corresponds to the part listed in the left sidebar.

> 1. In the new `encoder-1` panel, configure your component by updates the following attributes in the **CONFIGURE** field. This tells your encoder component to use a specific pin on a specific board (called `board-1` in the Viam app).

Reviewer: ```suggestion
1. In the new `encoder-1` panel, configure your component by updating the following attributes in the **CONFIGURE** field. This tells your encoder component to use a specific pin on a specific board (called `board-1` in the Viam app).
```

   - Select `board-1` from the dropdown under the "board" field.
   - Click the "Add pins" button to reveal the field for setting the encoder's interrupt pin
   - Set the "i" field to "22"
   ![configure encoder](assets/configureEncoder.webp)
1. Click **Save** in the top right to save and apply your configuration changes.
1. At the bottom of the `encoder-1` panel, expand the **TEST** section to ensure you have configured the encoder correctly. You should see the ticks count under "GetPosition" increase when pressing down and releasing the limit switch.
   ![test encoder](assets/testEncoder.webp)

### Configure your motor

1. In [the Viam app](https://app.viam.com/fleet/locations), find the **CONFIGURE** tab.
1. Click the **+** icon in the left-hand menu and select **Component**.
1. Select `motor`, and find the `gpio` module. This adds the module for controlling a motor with a board's GPIO pins. Leave the default name `motor-1` for now.
1. Notice adding this module adds the motor hardware component called `motor-1`. The collapsible card on the right corresponds to the part listed in the left sidebar.

> 1. In the new `motor-1` panel, configure your component by updates the following attributes in the **CONFIGURE** field. This tells your motor component to use specific pins and settings on a specific board (called `board-1` in the Viam app).

Reviewer: ```suggestion
1. In the new `motor-1` panel, configure your component by updating the following attributes in the **CONFIGURE** field. This tells your motor component to use specific pins and settings on a specific board (called `board-1` in the Viam app).
```

### Capture images from the webcam

1. In the `camera-1` panel, click the "+ Add method" button under the **Data capture** section.
1. For the "Method" dropdown, select "ReadImage", enter "0.25" in the "Frequency (hz)" field to capture an image every 4 seconds.
1. Make sure the data capture toggle is "On".
   ![add data capture method](./assets/configureCameraCapture.webp)
1. Click **Save** in the top right. This may take a moment to apply your configuration changes.
1. At the bottom of the `camera-1` panel, expand the **TEST** section to view the images as they're being captured.
   ![test camera](assets/testCamera.webp)
1. Click the "Toggle picture-in-picture" button to create a separate window of the camera view.
1. Go to the `motor-1` panel and expand the **TEST** section at the bottom. Use the controls to turn the motor and dispense the snacks into the bowl.
   ![dispense snacks with motor](assets/enableMotorControl.webp)
1. Once the bowl is full, empty the bowl back into the container and dispense more snacks. Go through this cycle a few times to ensure you have enough image data to train an accurate model.

> 1. When you feel you have enough images, you can close the picture-in-picture window of the camera and turn off data capture in the `camera-1` panel.

Reviewer: ```suggestion
1. When you feel you have enough images (with a minimum of 25), you can close the picture-in-picture window of the camera and turn off data capture in the `camera-1` panel.
```

Adding a minimum number of images to ensure a greater chance of their model working. Feel free to change the number.

   ![add ML model](assets/addMLModel.webp)
1. Notice this creates two new items in the left sidebar. The first is your new ML Model service called `mlmodel-1`, and the second is the `tflite_cpu` module from the Registry.
1. In the `mlmodel-1` panel, click "Select model" in the **Model** section, search for "snacks-refill" under the **My organization** tab, and select your trained model from the list.
   ![add model](assets/configureMLModel.webp)
1. Click **Save** in the top right to save and apply your configuration changes.

### Add a vision service

1. In the Viam app, click the **+** icon in the left-hand menu and select **Service**, and then `vision`.
1. Search for a module called `mlmodel`. Then click **Add module**, and **Create** a new [Vision service](https://docs.viam.com/services/vision/) called `vision-1`.
   ![add module](assets/addVision.webp)
1. Notice adding this service adds the vision service called `vision-1`. From the **ML Model** section of the panel, select `mlmodel-1`.
   ![select ML model for vision service](assets/configureVision.webp)
1. **Save** your changes in the top right and wait a few moments for the configuration changes to take effect.

> 1. At the bottom of the `vision-1` panel, expand the **TEST** section to ensure you have configured the service properly and 

Reviewer: Sentence cuts off. Maybe meant this?

```suggestion
1. At the bottom of the `vision-1` panel, expand the **TEST** section to ensure you have configured the service properly and can see the classifications table and its label predictions.
```

   }
   ```
   ![configure refiller service](assets/configureRefiller.webp)
1. **Save** your changes in the top right and wait a few moments for the configuration changes to take effect.
1. Test out the service by emptying the snack bowl and watch the machine automatically dispense more snacks until the bowl is full.

<!-- ![machine in action](assets/smartSnacks.gif) -->

> aside negative
> **TROUBLESHOOTING**: Double check your service configuration. Look at the **LOGS** tab to see what might be going wrong.

<!-- ------------------------ -->
## Conclusion And Resources
Duration: 1

Congratulations! You've just transformed a simple food dispenser into a smart device that makes snacking both fun and mindful.
While this project serves up treats, it's also given you a taste of key hardware hacking concepts - from motor control to computer vision to hardware automation.

### What You Learned
- How to configure and test a device's components using Viam
- How to use modules from the Viam registry
- How to train a custom ML model for computer vision
- How to use a vision service for image classification

### Building advanced scenarios with Viam and the smart snack dispenser

> At this point, you have configured and tested your machine to recognize a snack food and automatically refill the bowl. Consider adding feature like:

Reviewer: ```suggestion
At this point, you have configured and tested your machine to recognize a snack food and automatically refill the bowl. Consider adding features like:
```
